python/neuroglancer/ingest/volumes/gcloudvolume.py

```diff
@@ -0,0 +1,199 @@
+    blob = self.__getInfoBlob()
+    blob.upload_from_string(json.dumps(self.info), 'application/json')
+
+  @property
+  def base_cloudpath(self):
+    return "gs://{}/{}/".format(lib.GCLOUD_BUCKET_NAME, self.dataset_name)
+
+  @property
+  def shape(self):
+    return self.mip_shape(self.mip)
+
+  def mip_shape(self, mip):
+    return Vec3(*self.info['scales'][mip]['size'])
+
+  @property
+  def layer_type(self):
+    return self.info['type']
+
+  @property
+  def data_type(self):
+    return self.info['data_type']
+
+  @property 
+  def resolution(self):
+    return self.mip_resolution(self.mip)
+
+  def mip_resolution(self, mip):
+    return Vec3(*self.info['scales'][mip]['resolution'])
+
+  @property
+  def downsample_ratio(self):
+    return self.resolution / self.mip_resolution(0)
+
+  @property
+  def underlying(self):
+    return self.mip_underlying(self.mip)
+
+  def mip_underlying(self, mip):
+    return Vec3(*self.info['scales'][mip]['chunk_sizes'][0])
+
+  def __getitem__(self, slices):
+    while len(slices) < 3:
+      slices.append( slice() )
+
+    maxsize = Vec3(*self.info['scales'][self.mip]['size'])
+
+    slices = list(slices)
+    for index, slc in enumerate(slices):
+      if isinstance(slc, int) or isinstance(slc, float) or isinstance(slc, long):
+        slices[index] = slice(slc, slc+1, 1)
+      else:
+        start = 0 if slc.start is None else clamp(slc.start, 0, maxsize[index])
+        end = maxsize[index] if slc.stop is None else clamp(slc.stop, 0, maxsize[index])
+        step = 1 if slc.step is None else slc.step
+
+        slices[index] =  slice(start, end, step)
+
+    minpt = Vec3(*[ x.start for x in slices ]) * self.downsample_ratio
+    maxpt = Vec3(*[ x.stop for x in slices ]) * self.downsample_ratio
+    steps = Vec3(*[ x.step for x in slices ])
+
+    savedir = os.path.join(lib.COMMON_STAGING_DIR, 'gcloud', self.dataset_name, self.layer, str(self.mip))
+
+    return self.cutout(
+      xmin=minpt.x, xmax=maxpt.x, xstep=steps.x,
+      ymin=minpt.y, ymax=maxpt.y, ystep=steps.y,
+      zmin=minpt.z, zmax=maxpt.z, zstep=steps.z,
+      savedir=( savedir if self.cacheFiles else None ),
+    )
+
+  def cutout(self, xmin, xmax, ymin, ymax, zmin, zmax, xstep=1, ystep=1, zstep=1, savedir=None):
+    
+    try:
+      requested_mip_level = self.info['scales'][self.mip]
+    except IndexError:
+      raise Exception("{} mip level has not been generated. Max: {}".format(self.mip, len(self.info['scales']) - 1))
+
+    requested_bbox = Bbox(Vec3(xmin, ymin, zmin), Vec3(xmax, ymax, zmax)) / self.downsample_ratio
+    volume_bbox = Bbox.from_vec(Vec3(*requested_mip_level['size'])) # volume size in voxels
+    
+    realized_bbox = requested_bbox.fitToChunkSize(self.underlying)
+    realized_bbox = Bbox.clamp(requested_bbox, volume_bbox)
+    
+    cloudpaths = self.__cloudpaths(realized_bbox, volume_bbox, requested_mip_level['key'], self.underlying)
+    encoding = requested_mip_level['encoding'] # e.g. jpeg, raw
```

Comment:
I would move this line down to the if statement because there is went it get firstly used.